### DISCUSSION TITLE
need help!

### DISCUSSION THREAD

**Discussion by lbk861** [fltk-rs/fltk-rs#1546] [Q&A] [answered]
I encounter some questions which may be bugs? I don't know.
Below is my code. When I run the app and click the Hello button, everything is ok. Alert window displays correctly. But when I click switch button to switch to another window. Click wrong button, the app is stucked .
![image](https://github.com/fltk-rs/fltk-rs/assets/147920138/a9d62d3d-e841-4036-b7ad-098647cada3a)
![image](https://github.com/fltk-rs/fltk-rs/assets/147920138/dd942514-cdfa-42dd-8f67-648c4cf81ed8)

![image](https://github.com/fltk-rs/fltk-rs/assets/147920138/1d33f71e-a311-47aa-bac1-9d88d2801eb1)

```rs
#![cfg_attr(
    all(target_os = "windows", not(debug_assertions),),
    windows_subsystem = "windows"
)]

use fltk::app::App;
use fltk::button::Button;
use fltk::frame::Frame;
use fltk::group::Group;
use fltk::prelude::*;
use fltk::window::{DoubleWindow, Window};

fn main() {
    let app = App::default();
    let win_width = 500;
    let win_height = 300;
    let mut win = DoubleWindow::new(0, 0, win_width, win_height, "Test")
        .with_id("main_win")
        .center_screen();

    let group = Group::new(105, 0, 395, 270, None).with_id("group");
    let mut f = Button::default()
        .with_size(100, 40)
        .with_label("Hello")
        .center_of(&group);
    group.end();
    let mut btn = Button::default().with_size(100, 45).with_label("switch");

    win.show();

    f.set_callback(|_| {
        alert("world");
    });

    btn.set_callback(move |_| {
        let mut group: Group = fltk::app::widget_from_id("group").unwrap();
        group.clear();
        group.begin();
        view();
        group.end();
        fltk::app::awake();
        fltk::app::redraw();
    });

    app.run().unwrap();
}
fn view() {
    let group = Group::new(105, 0, 395, 270, None);
    let mut btn = Button::default()
        .with_label("wrong")
        .with_size(100, 40)
        .center_of(&group);

    btn.set_callback(move |_| {
        alert("ok");
    });

    group.end();
}
pub fn alert(content: &'static str) {
    let mut win = Window::default()
        .with_size(200, 100)
        .with_label("Alert")
        .center_screen();

    let f = Frame::new(0, 10, 80, 30, content).center_x(&win);

    let mut ok = Button::new(0, 0, 50, 30, "Ok")
        .below_of(&f, 25)
        .center_x(&win);

    win.make_modal(true);
    win.show();

    ok.set_callback({
        move |_| {
            win.hide();
        }
    });
}

```

**Comment by MoAlyousef** [comment, marked as answer]
Can you try adding `Group::set_current(None::<&Group>);` in your alert function:
```rust
pub fn alert(content: &'static str) {
    Group::set_current(None::<&Group>);
    // rest of code
}
```
FLTK uses some heuristics to determine whether a window is a child of another window. Apparently here it can't. You can either use the above described function, or create an alert window in your main, hide it, then show it when needed.

**Reply by lbk861** [reply]
Many thanks, It works. And I'm new to gui and fltk. I don't understand heuristics, is there anything about that let me learn ?

**Reply by MoAlyousef** [reply]
You're welcome. Since FLTK allows embedding windows within other windows, spawning a window while a group widget is being populated might cause the window to be embedded in the group widget. 
For other widgets, it's not a problem, but I did notice that it occurs with windows. 
As far as fltk-rs learning materials, there's the fltk-rs [book](https://fltk-rs.github.io/fltk-book/) that you can use, in addition to the [documentation](https://docs.rs/fltk) itself.
The definitive guide would be the FLTK book itself which can be found here:
https://www.fltk.org/doc-1.4/fltk.pdf
But the code samples are in C++.
